### PR TITLE
wal: record mark when snapshotting

### PR DESCRIPTION
**Pull request by yichengq**
{Index, Term} pair must be saved to WAL before loading it. I do strict checking now to ensure safety.

### CODE REVIEW

File: wal/walpb/record.proto [hunk 1]
```diff
@@ -12,3 +12,8 @@ message Record {
 	required uint32 crc  = 2 [(gogoproto.nullable) = false];
 	optional bytes data  = 3;
 }
+
+message Mark {
```
Comment (xiang90) [inline wal/walpb/record.proto]: Mark -> Snapshot

File: wal/wal.go [hunk 2]
```diff
@@ -47,6 +48,8 @@ var (
 	ErrMetadataConflict = errors.New("wal: conflicting metadata found")
 	ErrFileNotFound     = errors.New("wal: file not found")
 	ErrCRCMismatch      = errors.New("wal: crc mismatch")
+	ErrMarkMismatch     = errors.New("wal: mark mismatch")
```
Comment (xiang90) [inline wal/wal.go]: snapshot mismatch

File: wal/wal.go [hunk 3]
```diff
@@ -47,6 +48,8 @@ var (
...
+	ErrMarkNotFound     = errors.New("wal: mark not found")
```
Comment (xiang90) [inline wal/wal.go]: snapshot not found

File: wal/wal.go [hunk 4]
```diff
@@ -206,12 +209,13 @@ func (w *WAL) ReadAll() (metadata []byte, state raftpb.HardState, ents []raftpb.
 	rec := &walpb.Record{}
 	decoder := w.decoder
 
+	var match bool
```
Comment (xiang90) [inline wal/wal.go]: please add todos to readall. i think we need to detect more errors and loosen the checking of match.

File: wal/wal.go [hunk 5]
```diff
@@ -60,8 +63,8 @@ type WAL struct {
 	metadata []byte           // metadata recorded at the head of each WAL
 	state    raftpb.HardState // hardstate recorded at the head of WAL
 
-	ri      uint64   // index of entry to start reading
-	decoder *decoder // decoder to decode records
+	mark    walpb.Mark // mark to start reading
```
Comment (xiang90) [inline wal/wal.go]: lastSnap

File: wal/wal.go [hunk 6]
```diff
@@ -374,6 +396,19 @@ func (w *WAL) Save(st raftpb.HardState, ents []raftpb.Entry) error {
 	return w.sync()
 }
 
+func (w *WAL) SaveSnapshot(e walpb.Snapshot) error {
+	b := pbutil.MustMarshal(&e)
+	rec := &walpb.Record{Type: snapshotType, Data: b}
+	if err := w.encoder.encode(rec); err != nil {
+		return err
+	}
+	// update enti only when mark is ahead of last index
```
Comment (xiang90) [inline wal/wal.go]: mark -> snap

File: wal/walpb/record.pb.go [hunk 7]
```diff
@@ -10,6 +10,7 @@
 
 	It has these top-level messages:
 		Record
+		Mark
```
Comment (xiang90) [inline wal/walpb/record.pb.go]: Mark -> Snapshot?

**Comment by yichengq** [comment]
eagle eye! refreshed.

**Comment by xiang90** [comment]
@yichengq The code lgtm. we also need to update the doc for wal.

**Comment by xiang90** [comment]
lgtm after fixing the doc.

**Comment by yichengq** [comment]
@xiang90 updated.

**Comment by xiang90** [comment]
lgtm

**Pull request merged**